arch-riscv: Add senvcfg CSR

erin-le opened the pull request:
This commit adds the senvcfg control status register, which fixes the 6.11.3 kernel crash documented in issue 1674. I have not added a bitfield and its implementation in isa.cc only uses setMiscRegNoEffect, so this implementation is missing some components.

erin-le committed:
arch-riscv: Add senvcfg CSR

This commit adds the senvcfg CSR, which fixes the 6.11.3 kernel
crash documented in issue 1674. I have not added a bitfield and
its implementation in isa.cc only uses setMiscRegNoEffect, so
this implementation is likely missing some critical components.

BobbyRBruce reviewed (dismissed):
I'm very impressed you figured this out.

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -779,6 +780,11 @@ ISA::setMiscReg(RegIndex idx, RegVal val)
             break;
+          case MISCREG_SENVCFG:
+            {
+                setMiscRegNoEffect(idx, val + 1);
```
We should probably make this behave correctly before committing. I'll see if I have time today to look at it.

powerjg commented:
Specific things that should change before merging:
1. We should only support writing FIOM, not the other bitfileds. If the other bitfileds (i.e., the bitfields that the manual says "will be furnished by the forthcoming extension...") are written we should panic with an appropriate error so we can fix it later. We want to panic here so that in the future if the kernel (or some other application) writes these bits we know how to implement it. Or, maybe we'll chose to ignore it, but let's have our future selves make that decision.
2. This should probably not be a `setMiscRegNoEffect` but we should use the normal `setMiscReg` function (I believe). 
3. We should make sure to ignore the writes to any of the bits marked `WPRI`. Don't panic in this case, but ignore the writes.

I don't believe we need to change the implementation of atomics or fence instructions because I believe they already have the same ordering for memory and I/O space implied. I.e., gem5 takes an conservative approach. However, we *might* need to read the `senvcfg` in the fence and atomic instruction implementations. I think that we can skip that for now, but it is a true dependence that may need to be enforced.

Thanks for picking this up Erin, and great job getting started on it! Everything that I said above you couldn't figure out without a lot more background on RISC-V and gem5's implementation. Again, great job!

Edit: I was wrong about what to do with `wpri` fields.

erin-le commented:
@rogerchang23424 Feedback on this would be greatly appreciated!

erin-le commented:
@powerjg: I implemented the bitfields according to the RISCV spec and added panics/warnings based on what you said above, but it looks like the instruction that accesses senvcfg tries to write to the CBZE bitfield, which causes a panic. Should the panic be a warning instead?

powerjg commented:
> it looks like the instruction that accesses senvcfg tries to write to the CBZE bitfield

Then the panic worked as expected! Let's figure out how to implement that :)

Here's the definition: https://github.com/riscv/riscv-CMOs/blob/8e8a8e01e1e2761579eaaa0f1c0f0c1efac9eddb/cmobase/csr_state.adoc#L51

erin-le committed:
arch-riscv: Implement behavior for senvcfg register

This commit adds behavior for writes to the senvcfg register.
It also implements the CBIE, CBCFE, and CBZE bitfields of
senvcfg.

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -779,6 +780,31 @@ ISA::setMiscReg(RegIndex idx, RegVal val)
+                }
+
+                if ((wpri_mask & val) != 0) {
+                    warn("Ignoring write to WPRI bit(s) in senvcfg CSR.\n"
```
Do we print a warning in other cases of wrpi writes? If so, we should keep this. If not, we should not have this warning.

erin-le replied:
It doesn't look like there's a warning in other cases, so I've removed this

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -779,6 +780,31 @@ ISA::setMiscReg(RegIndex idx, RegVal val)
+                    warn("Ignoring write to WPRI bit(s) in senvcfg CSR.\n"
+                    "The attempted write was:\n %" PRIu64 "\n", val);
+                } else {
+                    setMiscRegNoEffect(idx, val);
```
Can you simply call `setMiscReg`? I think we want the effect.

erin-le replied:
I've left this call as setMiscRegNoEffect, as this part of the code is already in setMiscReg. It's just used as a setter function here with the code above being the effect. I think that's how the other registers also use setMiscRegNoEffect.

I've also changed the implementation so `val` excludes the WPRI bits.

powerjg reviewed (commented):
One more small thing. Great job!!

erin-le committed:
arch-riscv: Remove warning message in senvcfg setMiscReg

Previously, a warning would be printed when the WPRI bits
in the senvcfg register were written to. The other registers
do not print warnings for this, so the warning is being
removed.

rogerchang23424 commented on src/arch/riscv/regs/misc.hh:
```diff
@@ -183,6 +183,7 @@ enum MiscRegIndex
     MISCREG_SCAUSE,
     MISCREG_STVAL,
     MISCREG_SATP,
+    MISCREG_SENVCFG,
```
We forgot adding extra 0 for `MISCREG_SENVCFG` in `util/cpt_upgrader`. We should upgrade the checkpoint to avoid inconsistent miscreg size between old and new checkpoint version.
